src/jupyterhub_config/hub_singleuser.py

```diff
@@ -0,0 +1,143 @@
+import fcntl
+import grp
+import logging
+import os
+import pwd
+import subprocess
+import tempfile
+from pathlib import Path
+
+
+class SingleUserEnvManager:
+    """
+    Manages the environment and user-specific settings for a single user in a JupyterHub server.
+    """
+
+    def __init__(self,
+                 username,
+                 groupname: str = 'jupyterhub', ):
+        """
+        Initializes the environment manager for the specified user.
+        :param username: The name of the user for whom the environment is set up.
+        :param groupname: The name of the OS system group to which the user belongs.
+        """
+        self.username = username  # OS system username
+        self.groupname = groupname  # OS system group
+
+        # Inherit the current environment variables
+        self.environment = dict(os.environ)
+
+        self.global_home = Path(os.environ['JUPYTERHUB_USER_HOME'])
+        self.user_dir = self.global_home / username
+        self.log = self._init_logger()
+
+    def setup_environment(self):
+        """
+        Set up the user's environment by ensuring the system user exists, and the workspace
+        permissions are correctly configured.
+        """
+        self.log.info(f"Setting up environment for user: {self.username}")
+        self._ensure_system_user()
+        self._ensure_workspace_permission()
+
+    def start_single_user_server(self):
+        """
+        Starts the single-user Jupyter server for the user, inheriting the configured environment.
+        """
+        self.log.info(f'Starting single-user server for {self.username} with environment: {self.environment}')
+        env_vars = [f'{key}={value}' for key, value in self.environment.items()]
+
+        cmd = ['sudo', '-E', '-u', self.username, 'env'] + env_vars + [
+            os.path.join(os.environ['JUPYTERHUB_CONFIG_DIR'], 'spawn_notebook.sh')]
+
+        self.log.info(f'Executing command: {" ".join(cmd)}')
+        subprocess.run(cmd, check=True)
+
+    def _ensure_system_user(self):
```

Comment:
mostly copied from [customer_spawner.py](https://github.com/kbase/cdm-jupyterhub/blob/main/src/jupyterhub_config/custom_spawner.py#L62). But using `pwd` and `grp` to check existing user and group.